src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -293,6 +289,30 @@ ClientRequest::do_process(
     return reply_op_error(pg, -ENOENT);
   }
 
+  if (!pg->is_primary()) {
+    // primary can handle both normal ops and balanced reads
+    if (is_misdirected(*pg)) {
+      logger().trace("do_process: dropping misdirected op");
+      return seastar::now();
+    } else if (const hobject_t& hoid = m->get_hobj();
+               !pg->get_peering_state().can_serve_replica_read(hoid)) {
+      logger().debug("{}: unstable write on replica, "
+	             "bouncing to primary",
+                     __func__);
+      return reply_op_error(pg, -EAGAIN);
+    } else {
+      logger().debug("{}: serving replica read on oid {}",
+                     __func__, m->get_hobj());
+    }
+  }
+  return _do_process(ihref, pg, obc);
```

Comment:
This doesn't seem helpful.  Why not leave it as a single method?

Reply:
Unified